Source/Core/Core/Lua/LuaEventCallbackClasses/LuaWheneverCallbackClass.cpp

```diff
@@ -0,0 +1,127 @@
+#include "Core/Lua/LuaEventCallbackClasses/LuaWheneverCallbackClass.h"
+
+#include <array>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <random>
+#include <thread>
+#include <vector>
+
+#include "Core/Lua/LuaHelperClasses/LuaColonCheck.h"
+#include "Core/Lua/LuaVersionResolver.h"
+
+namespace Lua::LuaWheneverCallback
+{
+
+struct LuaThreadToFunctionReference
+{
+  lua_State* lua_thread;
+  int function_reference;
+};
+
+class LuaWhenever
+{
+public:
+  inline LuaWhenever() {}
+};
+
+static std::vector<LuaThreadToFunctionReference> list_of_functions_to_run;
+static bool list_of_functions_is_empty = true;
+static std::unique_ptr<LuaWhenever> instance = nullptr;
+static std::thread lua_whenever_callback_runner_thread;
+static std::mutex list_of_functions_lock;
+static std::condition_variable list_of_functions_condition_variable;
+static std::mt19937_64 rng_generator;
+static int temp_int;
+
+LuaWhenever* GetInstance()
+{
+  if (instance == nullptr)
+    instance = std::make_unique<LuaWhenever>(LuaWhenever());
+  return instance.get();
+}
+
+void InitLuaWheneverCallbackFunctions(lua_State* lua_state, const std::string& lua_api_version)
```

Comment:
You should be able to make a generic class that is parameterised for callback type.

Reply:
@phire Are you talking about making a template class?